cmake/BsoncxxUtil.cmake

```diff
@@ -10,17 +10,147 @@ function(bsoncxx_add_library TARGET OUTPUT_NAME LINK_TYPE)
         ${bsoncxx_sources}
     )
 
+    # Full ABI tag string to append to library output name.
+    # The value is determined at generator-time when using a multi-config generator.
+    # Otherwise, the value is determined at configure-time.
+    set(abi_tag "")
+
+    # ABI tag and properties.
+    if(1)
+        # Enforce ABI compatibility in dependent targets.
+        set_property(TARGET ${TARGET} APPEND PROPERTY COMPATIBLE_INTERFACE_STRING
+            BSONCXX_ABI_TAG_BUILD_TYPE
+            BSONCXX_ABI_TAG_MONGOC_LINK_TYPE
+            BSONCXX_ABI_TAG_POLYFILL_LIBRARY
+        )
+
+        # ABI version number. Only necessary for shared library targets.
+        if(LINK_TYPE STREQUAL "SHARED")
+            set(soversion _noabi)
+            set_target_properties(${TARGET} PROPERTIES SOVERSION ${soversion})
+            string(APPEND abi_tag "-v${soversion}")
+        endif()
+
+        # Build type:
+        # - 'd' for debug.
+        # - 'r' for release (including RelWithDebInfo and MinSizeRel).
+        # - 'u' for unknown.
+        if(1)
+            if(isMultiConfig)
+                set(build_type $<IF:$<CONFIG:Debug>,d,$<IF:$<OR:$<CONFIG:Release>,$<CONFIG:RelWithDebInfo>,$<CONFIG:MinSizeRel>>,r,u>>)
+            else()
+                if(CMAKE_BUILD_TYPE STREQUAL "Debug")
+                    set(build_type "d")
+                elseif(CMAKE_BUILD_TYPE STREQUAL "Release" OR CMAKE_BUILD_TYPE STREQUAL "RelWithDebInfo" OR CMAKE_BUILD_TYPE STREQUAL "MinSizeRel")
+                    set(build_type "r")
+                else()
+                    set(build_type "u")
+                endif()
+            endif()
+
+            set_target_properties(${TARGET} PROPERTIES INTERFACE_BSONCXX_ABI_TAG_BUILD_TYPE ${build_type})
+
+            string(APPEND abi_tag "-${build_type}")
+        endif()
+
+        # Link type with libmongoc.
+        # - 'h' for shared.
+        # - 't' for static.
+        if(1)
+            if(BSONCXX_LINK_WITH_STATIC_MONGOC)
+                set(mongoc_link_type "t")
+            else()
+                set(mongoc_link_type "h")
+            endif()
+
+            set_target_properties(${TARGET} PROPERTIES INTERFACE_BSONCXX_ABI_TAG_MONGOC_LINK_TYPE ${mongoc_link_type})
+
+            string(APPEND abi_tag "${mongoc_link_type}")
+        endif()
+
+        # Library used for C++17 polyfills:
+        # - 'm' for mnmlstc/core.
+        # - 'b' for Boost.
+        # - 'x' for experimental standard library.
+        # - 'i' for bsoncxx (re)implementations.
+        # - 's' for standard library (no polyfill).
+        if(1)
+            if(BSONCXX_POLY_USE_MNMLSTC)
+                set(polyfill "m")
+            elseif(BSONCXX_POLY_USE_BOOST)
+                set(polyfill "b")
+            elseif(BSONCXX_POLY_USE_STD_EXPERIMENTAL)
+                set(polyfill "x")
+            elseif(BSONCXX_POLY_USE_STD)
+                set(polyfill "s")
+            else()
+                set(polyfill "u")
```

Comment:
"u" was meant to represent "unknown", similar to the build type. We can consider removing the "u" values entirely (both for the polyfill parameter and the build type parameter) and treat the unknown case as a configuration error if preferable.

Reply:
The point I was making was the comment lists 'mbxis' as the options, but the logic has branches for 'mbxus'. It wasn't clear to me if your branch for 'u' was meant to be the same thing as "- 'i' for bsoncxx (re)implementations." It would seem that either the 'i' entry should be remove from the comment (and replaced with a comment for 'u') or that a branch should be added to the conditional so that the 'i' case is handled. Though, I'm not sure which polyfill option would correspond to 'i', off the top of my head.

Sorry for not being more clear in my first comment.

Reply:
Oh, right, sorry. This is because the specifics of how bsoncxx implementations are chosen as polyfill is not yet determined; that will be addressed by CXX-2796. Currently, pending 3.10 release changes unconditionally use the bsoncxx implementation without any regards to availability of C++17 features. I (perhaps prematurely) reserved `i` for when we have only `i` or `s` (dropping `m`, `b`, and `x` in CXX-2797).

I decided to drop `u` for simplicity and added a "branch stub" for the `i` case that we can enable/update in the future.